### DISCUSSION TITLE
Making a Socket io singleton class to be used in any file

### DISCUSSION THREAD

**Discussion by hmzakhalid** [socketio/socket.io#4173] [Q&A] [answered]
I need access to the `io` module in multiple files in my project so I can emit some messages. So for that I decided to create a singleton class to handle the `io` object so i can import the class anywhere and use the `io` object.

When I comment out all the middleware, it works. But I don't want that, I want the connection to be only made when the user is logged in.

The `io.on('connect')` wont console.log anything on a new connection. What is the problem here?

here's my `socketClass.js`
```javascript
const session = require('express-session');
const { Server } = require("socket.io");
const passport = require('passport');


class SocketManager{
    constructor() {
        this.io = null;
    }

    start(server) {
        this.io = new Server(server);

        // convert a connect middleware to a Socket.IO middleware
        const wrap = middleware => (socket, next) => middleware(socket.request, {}, next);

        const sessionMiddleware = session({
            secret: 'MySecrectKey',
            resave: true,
            saveUninitialized: true
        });
        this.io.use(wrap(sessionMiddleware));
        this.io.use(wrap(passport.initialize()));
        this.io.use(wrap(passport.session()));

        this.io.use((socket, next) => {
            if (socket.request.user) {
                next();
            } else {
                next(new Error('unauthorized'))
            }
        });
    }

    get() {
        return this.io;
    }

}

module.exports = new SocketManager();
```

Here's my `app.js`
```javascript
...
const SocketManager = require('./socketClass');
const server = http.createServer(app);
SocketManager.start(server);
const io = SocketManager.get();

io.on('connect', async (socket) => {
    console.log(`new connection ${socket.id}`);
    socket.on('whoami', (cb) => {
        cb(socket.request.user ? socket.request.user.username : '');
    });
});
```

**Comment by darrachequesne** [comment, marked as answer]
The `connect` event is not emitted, because the connection is denied with `next(new Error('unauthorized'))`. The `socket.request.session` is defined, but passport is not able to fetch the user from it. I think you are missing an authentication strategy, like http://www.passportjs.org/docs/username-password/.

You will find a complete example with Passport here: https://github.com/socketio/socket.io/tree/master/examples/passport-example

**Reply by hmzakhalid** [reply]
Hey, Thanks for replying. I figured it out this morning and fixed it!
